arangod/RocksDBEngine/RocksDBIterators.cpp

```diff
@@ -35,315 +34,354 @@
-
-  // note: this is always a forward iterator
-  return _mustCheckBounds && _cmp->Compare(_iterator->key(), _upperBound) > 0;
-}
-
-bool RocksDBAllIndexIterator::nextImpl(LocalDocumentIdCallback const& cb,
-                                       size_t limit) {
-  ensureIterator();
-  TRI_ASSERT(_trx->state()->isRunning());
-  TRI_ASSERT(_iterator != nullptr);
-
-  if (limit == 0 || ADB_UNLIKELY(!_iterator->Valid()) || outOfRange()) {
-    // No limit no data, or we are actually done. The last call should have
-    // returned false
-    TRI_ASSERT(limit > 0);  // Someone called with limit == 0. Api broken
-    // validate that Iterator is in a good shape and hasn't failed
-    arangodb::rocksutils::checkIteratorStatus(_iterator.get());
-    return false;
   }
 
-  TRI_ASSERT(limit > 0);
+  std::string_view typeName() const noexcept override {
+    return "all-index-iterator";
+  }
 
-  do {
-    TRI_ASSERT(_bounds.objectId() == RocksDBKey::objectId(_iterator->key()));
+  /// @brief index does not support rearming
+  bool canRearm() const override { return false; }
 
-    cb(RocksDBKey::documentId(_iterator->key()));
-    --limit;
-    _iterator->Next();
+  bool nextImpl(LocalDocumentIdCallback const& cb, uint64_t limit) override {
+    ensureIterator();
+    TRI_ASSERT(_trx->state()->isRunning());
+    TRI_ASSERT(_iterator != nullptr);
 
-    if (ADB_UNLIKELY(!_iterator->Valid())) {
+    if (!_iterator->Valid() || outOfRange() || ADB_UNLIKELY(limit == 0)) {
+      // No limit no data, or we are actually done. The last call should have
+      // returned false
+      TRI_ASSERT(limit > 0);  // Someone called with limit == 0. Api broken
       // validate that Iterator is in a good shape and hasn't failed
       arangodb::rocksutils::checkIteratorStatus(_iterator.get());
       return false;
-    } else if (outOfRange()) {
-      return false;
     }
 
-    if (limit == 0) {
-      return true;
-    }
-  } while (true);
-}
+    TRI_ASSERT(limit > 0);
 
-bool RocksDBAllIndexIterator::nextDocumentImpl(
-    IndexIterator::DocumentCallback const& cb, size_t limit) {
-  ensureIterator();
-  TRI_ASSERT(_trx->state()->isRunning());
-  TRI_ASSERT(_iterator != nullptr);
+    do {
+      TRI_ASSERT(_bounds.objectId() == RocksDBKey::objectId(_iterator->key()));
 
-  if (limit == 0 || !_iterator->Valid() || outOfRange()) {
-    // No limit no data, or we are actually done. The last call should have
-    // returned false
-    TRI_ASSERT(limit > 0);  // Someone called with limit == 0. Api broken
-    // validate that Iterator is in a good shape and hasn't failed
-    arangodb::rocksutils::checkIteratorStatus(_iterator.get());
-    return false;
-  }
+      cb(RocksDBKey::documentId(_iterator->key()));
+      _iterator->Next();
 
-  TRI_ASSERT(limit > 0);
+      if (ADB_UNLIKELY(!_iterator->Valid())) {
+        // validate that Iterator is in a good shape and hasn't failed
+        arangodb::rocksutils::checkIteratorStatus(_iterator.get());
```

Comment:
I also suggest to change checkIteratorStatus to work with reference not pointer, because it implementation can't work with nullptr